src/Command/Factory.php

```diff
@@ -109,10 +96,12 @@ public function createCommand($commandID, array $arguments = array())
      */
     public function defineCommand($commandID, $class)
     {
-        $reflection = new \ReflectionClass($class);
+        if ($class !== null) {
```

Comment:
maybe the docblock needs adjusting

I would recommend adding phpstan up to level 5 (minimum for packages usually for good quality).

Reply:
> Its type is `string`, where `null` comes from?

Indeed the docblock does not fully reflect the current behaviour of this method.

Here `NULL` is used to "undefine" the class handler for the specified command ID ([link to tests](https://github.com/predis/predis/blob/c26909699b6ad9bf17afb2becb619a59101026dd/tests/Predis/Command/RedisFactoryTest.php#L95-L128) just for reference), basically it says to the command factory to forget about any known handler for that command. By default `Predis\Command\Factory` uses a convention-over-configuration approach (different to command profiles in v1.1.) where `$commandID` is translated to the fully-qualified class name `Predis\Command\Redis\$commandID` when invoking `Predis\Command\Factory::createCommand()`. This is just the default behaviour as it can be overridden on a command basis by using `Predis\Command\Factory::defineCommand()`.

When `NULL` is passed as the second argument to this command, `NULL` is stored as the value in the internal named array that explicitly maps command IDs to PHP classes (this also prevent the command factory to pick the class handler by autoloading `Predis\Command\Redis\$commandID`).

Admittedly using `NULL` like this is not good here, and by doing so we are also exposing an implementation detail of the class. It would make more sense to have a `Predis\Command\Factory::undefineCommand()` method and use it accordingly [here in `Predis\Configuration\Option\Commands`](https://github.com/predis/predis/blob/4b2b35202acbaf79c2f543fc0be1d3520acdbad3/src/Configuration/Option/Commands.php#L35-L43) (`commands` is a new client option replacing the old `profile` one).

I'll fix this separately.

Reply:
> I would recommend adding phpstan up to level 5 (minimum for packages usually for good quality).

Running phpstan against Predis is indeed something I'd like to try sooner or later, I'm just not sure how it would handle the mix used in the library with a (mostly) strongly-typed approach for the inner parts and an heavily loosely-typed approach for the front-facing stuff (`Predis\Client` and related interfaces) most commonly used by developers.

Reply:
It would work for both just fine IMO, as it only finds the violations of specs, param/return and annotations etc.